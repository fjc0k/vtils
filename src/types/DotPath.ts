// ref: https://github.com/ahejlsberg/tsconf2020-demos/blob/master/template/main.ts#L108
type SubKeys<T, K extends string | number> = K extends keyof T
  ? `${K}.${DotPath<T[K]>}`
  : never

export type DotPath<T> = object extends T
  ? string
  : T extends any[]
  ? '0' | SubKeys<T, 0>
  : T extends readonly any[]
  ? Extract<keyof T, `${number}`> | SubKeys<T, Extract<keyof T, `${number}`>>
  : T extends object
  ? Extract<keyof T, string> | SubKeys<T, Extract<keyof T, string>>
  : never

export type DotPathValue<T, Path extends string> = Path extends '0'
  ? // @ts-ignore
    T[0]
  : Path extends keyof T
  ? T[Path]
  : Path extends `${infer K}.${infer R}`
  ? K extends '0'
    ? DotPathValue<
        // @ts-ignore
        T[0],
        R
      >
    : K extends keyof T
    ? DotPathValue<T[K], R>
    : unknown
  : unknown
