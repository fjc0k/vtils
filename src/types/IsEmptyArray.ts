import { NonEmptyArray } from './NonEmptyArray.ts'

/**
 * 判断 `T` 是否是空数组。
 *
 * @public
 * @example
 * ```typescript
 * type X = IsEmptyArray<[]>
 * // => true
 * ```
 */
export type IsEmptyArray<T> = T extends any[]
  ? T extends NonEmptyArray<any>
    ? false
    : true
  : false
