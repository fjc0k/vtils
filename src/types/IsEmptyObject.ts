import { IsNever } from './IsNever.ts'

/**
 * 判断 `T` 是否是空对象。
 *
 * @public
 * @example
 * ```typescript
 * type X = IsEmptyObject<{}>
 * // => true
 * ```
 */
export type IsEmptyObject<T> = T extends Object ? IsNever<keyof T> : false
