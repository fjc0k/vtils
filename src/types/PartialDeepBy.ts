// https://github.com/sindresorhus/type-fest/blob/main/source/set-required-deep.d.ts

import {
  IsAny,
  Paths,
  RequiredDeep,
  SetOptional,
  SimplifyDeep,
  UnionToTuple,
  UnknownArray,
} from 'type-fest'
import { NonRecursiveType, StringToNumber } from 'type-fest/source/internal'

export type PartialDeepBy<
  BaseType,
  KeyPaths extends Paths<BaseType>,
> = IsAny<KeyPaths> extends true
  ? SimplifyDeep<RequiredDeep<BaseType>>
  : PartialDeepByHelper<BaseType, UnionToTuple<KeyPaths>>

type PartialDeepByHelper<
  BaseType,
  KeyPathsTuple extends UnknownArray,
> = KeyPathsTuple extends [infer KeyPath, ...infer RestPaths]
  ? PartialDeepByHelper<PartialDeepBySinglePath<BaseType, KeyPath>, RestPaths>
  : BaseType

type PartialDeepBySinglePath<BaseType, KeyPath> =
  BaseType extends NonRecursiveType
    ? BaseType
    : KeyPath extends `${infer Property}.${infer RestPath}`
    ? {
        [Key in keyof BaseType]: Property extends `${Key & (string | number)}`
          ? PartialDeepBySinglePath<BaseType[Key], RestPath>
          : BaseType[Key]
      }
    : SetOptional<
        BaseType,
        (KeyPath | StringToNumber<KeyPath & string>) & keyof BaseType
      >
