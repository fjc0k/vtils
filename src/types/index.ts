/**
 * 类型工具库。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export type {
  LiteralUnion,
  AsyncReturnType,
  FixedLengthArray,
  PackageJson,
  TsConfigJson,
  JsonValue,
  JsonArray,
  JsonObject,
  CamelCase,
  SnakeCase,
  KebabCase,
  PascalCase,
  DelimiterCase,
  ScreamingSnakeCase as ConstantCase,
  Class,
  Asyncify,
  UnionToIntersection,
  Integer,
  NegativeInteger,
  NonNegativeInteger,
  Negative,
  NonNegative,
  Finite,
  PositiveInfinity,
  NegativeInfinity,
  Simplify,
  RequireAtLeastOne,
  RequireExactlyOne,
  RequireAllOrNone,
} from 'type-fest'

export type {
  AnyArray,
  ValueOf,
  ElementOf,
  AsyncOrSync,
  Buildable,
  Writable,
  Merge,
  Head,
  Tail,
  PickProperties as PickBy,
  OmitProperties as OmitBy,
  MarkOptional as PartialBy,
  MarkRequired as RequiredBy,
  StrictOmit as OmitStrict,
  DeepOmit as OmitDeep,
  DeepReadonly as ReadonlyDeep,
  DeepPartial as PartialDeep,
  DeepRequired as RequiredDeep,
  DeepWritable as WritableDeep,
  DeepNullable as NullableDeep,
  DeepNonNullable as NonNullableDeep,
  ReadonlyKeys,
  WritableKeys,
  OptionalKeys,
  RequiredKeys,
} from 'ts-essentials'

// @index(['./**/*.ts', '!./**/*.test.*', '!**/__*'], f => `export * from '${f.path}'`)
export * from './AnyAsyncFunction'
export * from './AnyFunction'
export * from './AnyObject'
export * from './Defined'
export * from './DotPath'
export * from './FirstParameter'
export * from './IsAny'
export * from './IsEmptyArray'
export * from './IsEmptyObject'
export * from './IsNever'
export * from './NonEmptyArray'
export * from './Nullable'
export * from './OneOrMore'
export * from './Path'
// @endindex
