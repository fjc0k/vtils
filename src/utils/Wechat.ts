import { EventBus, EventBusOffListener } from './EventBus'
import { loadResource, LoadResourceUrlType } from './loadResource'
import { mapValues, noop } from 'lodash-uni'

declare const wx: any

/**
 * 微信 JSSDK 支持的 API。
 *
 * @public
 */
export type WechatJsApi =
  | 'checkJsApi'
  | 'updateAppMessageShareData'
  | 'updateTimelineShareData'
  | 'onMenuShareTimeline'
  | 'onMenuShareAppMessage'
  | 'onMenuShareQQ'
  | 'onMenuShareQZone'
  | 'startRecord'
  | 'stopRecord'
  | 'onVoiceRecordEnd'
  | 'playVoice'
  | 'pauseVoice'
  | 'stopVoice'
  | 'onVoicePlayEnd'
  | 'uploadVoice'
  | 'downloadVoice'
  | 'chooseImage'
  | 'previewImage'
  | 'uploadImage'
  | 'downloadImage'
  | 'translateVoice'
  | 'getNetworkType'
  | 'openLocation'
  | 'getLocation'
  | 'hideOptionMenu'
  | 'showOptionMenu'
  | 'hideMenuItems'
  | 'showMenuItems'
  | 'hideAllNonBaseMenuItem'
  | 'showAllNonBaseMenuItem'
  | 'closeWindow'
  | 'scanQRCode'
  | 'chooseWXPay'
  | 'openProductSpecificView'
  | 'addCard'
  | 'chooseCard'
  | 'openCard'

/**
 * @public
 */
export interface WechatConfigParams {
  /**
   * 开启调试模式。
   *
   * 调用的所有api的返回值会在客户端 alert 出来，
   * 若要查看传入的参数，可以在 pc 端打开，
   * 参数信息会通过 log 打出，仅在 pc 端时才会打印。
   *
   * @default false
   */
  debug?: boolean

  /**
   * 公众号的唯一标识。
   */
  appId: string

  /**
   * 生成签名的时间戳。
   */
  timestamp: number | string

  /**
   * 生成签名的随机串。
   */
  nonceStr: string

  /**
   * 签名。
   */
  signature: string

  /**
   * 需要使用的JS接口列表。
   *
   * @default []
   */
  jsApiList?: WechatJsApi[]

  /**
   * 是否可分享。
   *
   * 设置为 `true` 将把分享系列接口自动加入 `jsApiList`。
   *
   * @default true
   */
  sharable?: boolean

  /**
   * 当全局变量 `wx` 不存在时，自动引入微信 JSSDK。
   *
   * - 设为 `false` 可禁止自动引入；
   * - 设为版本号表示引入特定版本的 JSSDK；
   * - 设为链接表示通过该链接引入 JSSDK。
   *
   * @default '1.4.0'
   */
  autoLoadJSSDK?: false | string
}

/**
 * @public
 */
export type WechatErrorCallback = (err: any) => void

/**
 * @public
 */
export interface WechatUpdateShareDataParams {
  /** 分享标题 */
  title?: string | (() => string)

  /** 分享描述 */
  desc?: string | (() => string)

  /** 分享链接，该链接域名或路径必须与当前页面对应的公众号 JS 安全域名一致 */
  link?: string | (() => string)

  /** 分享图标地址 */
  imgUrl?: string | (() => string)
}

/**
 * @public
 */
export interface WechatChooseImageParams {
  /**
   * 选择图片数量。
   *
   * @default 9
   */
  count?: number

  /**
   * 图片质量，可以指定是原图还是压缩图。
   *
   * @default ['original', 'compressed']
   */
  sizeType?: Array<'original' | 'compressed'>

  /**
   * 选择来源，可以指定是相册还是相机。
   *
   * @default ['album', 'camera']
   */
  sourceType?: Array<'album' | 'camera'>
}

/**
 * @public
 */
export interface WechatPreviewImageParams {
  /**
   * 当前显示图片的链接。
   *
   * @default urls[0]
   */
  current?: string

  /**
   * 需要预览的图片链接列表。
   */
  urls: string[]
}

/**
 * @public
 */
export interface WechatUploadImageParams {
  /**
   * 需要上传的图片的本地 ID，由 chooseImage 接口获得。
   */
  localId: string

  /**
   * 是否显示进度提示。
   *
   * @default false
   */
  isShowProgressTips?: boolean
}

/**
 * @public
 */
export interface WechatGetLocationParams {
  /**
   * 默认为 wgs84 的 gps 坐标，如果要返回直接给 openLocation 用的火星坐标，可传入 gcj02。
   */
  type?: 'wgs84' | 'gcj02'
}

/**
 * @public
 */
export interface WechatGetLocationResult {
  /**
   * 纬度，浮点数，范围为90 ~ -90。
   */
  latitude: number

  /**
   * 经度，浮点数，范围为180 ~ -180。
   */
  longitude: number

  /**
   * 速度，以米/每秒计。
   */
  speed: number

  /**
   * 位置精度。
   */
  accuracy: number
}

/**
 * @public
 */
export interface WechatOpenLocationParams {
  /**
   * 纬度，浮点数，范围为 90 ~ -90。
   */
  latitude: number

  /**
   * 经度，浮点数，范围为 180 ~ -180。
   */
  longitude: number

  /**
   * 位置名。
   */
  name: string

  /**
   * 地址详情说明。
   */
  address?: string

  /**
   * 地图缩放级别，整形值，范围从 1 ~ 28。
   *
   * @default 28
   */
  scale?: number

  /**
   * 在查看位置界面底部显示的超链接，可点击跳转。
   */
  infoUrl?: string
}

export interface WechatRequestPaymentParams {
  /**
   * 支付签名时间戳。
   */
  timestamp: number

  /**
   * 支付签名随机串。
   */
  nonceStr: string

  /**
   * 统一支付接口返回的 prepay_id 参数值，如：`prepay_id=xxx`。
   */
  package: string

  /**
   * 签名方式。
   */
  signType: string

  /**
   * 支付签名。
   */
  paySign: string
}

/**
 * 微信内网页的非基础菜单列表。
 *
 * @public
 */
export type WechatNonBaseMenuItem =
  | 'menuItem:share:appMessage'
  | 'menuItem:share:timeline'
  | 'menuItem:share:qq'
  | 'menuItem:share:weiboApp'
  | 'menuItem:favorite'
  | 'menuItem:share:facebook'
  | 'menuItem:share:QZone'
  | 'menuItem:editTag'
  | 'menuItem:delete'
  | 'menuItem:copyUrl'
  | 'menuItem:originPage'
  | 'menuItem:readMode'
  | 'menuItem:openWithQQBrowser'
  | 'menuItem:openWithSafari'
  | 'menuItem:share:email'
  | 'menuItem:share:brand'

/**
 * 微信 JSSDK 支持的分享 API 列表。
 */
const shareJsApiList: WechatJsApi[] = [
  'updateAppMessageShareData',
  'updateTimelineShareData',
  'onMenuShareAppMessage',
  'onMenuShareTimeline',
  'onMenuShareQQ',
  'onMenuShareQZone',
]

/**
 * 对微信 JSSDK 的封装。
 *
 * @public
 * @example
 * ```typescript
 * const wechat = new Wechat()
 * getWechatConfigAsync().then(config => {
 *   wechat.config(config)
 * })
 * wechat.updateShareData({
 *   title: '分享标题',
 *   desc: '分享描述',
 *   link: '分享链接',
 *   imgUrl: '缩略图地址',
 * })
 * wechat.invoke('scanQRCode').then(res => {
 *   // => API 调用结果
 * })
 * ```
 */
export class Wechat {
  /**
   * 微信 JSSDK 是否已准备完成。
   */
  private ready = false

  /**
   * 消息巴士。
   */
  private bus = new EventBus<{
    /**
     * 微信 JSSDK 准备完成时触发。
     */
    ready: () => void
    /**
     * 调用微信 JSSDK 出错时时触发。
     */
    error: WechatErrorCallback
  }>()

  /**
   * 上一次设置分享时的参数。
   */
  private prevShareParams: WechatUpdateShareDataParams = {}

  /**
   * 注入微信 `JSSDK` 的权限验证配置参数。
   */
  public configParams: WechatConfigParams = {} as any

  /**
   * 构造函数。
   *
   * @param params 注入微信 `JSSDK` 的权限验证配置参数
   */
  constructor(params?: WechatConfigParams) {
    if (params) {
      this.config(params)
    }
  }

  /**
   * 注入微信 `JSSDK` 的权限验证配置。
   *
   * @param params 配置参数
   */
  config(params: WechatConfigParams = this.configParams) {
    this.configParams = params

    const config = () => {
      const sharable =
        typeof params.sharable === 'boolean' ? params.sharable : true
      wx.config({
        ...params,
        jsApiList: [
          ...(params.jsApiList || []),
          ...(sharable ? shareJsApiList : []),
        ],
      })
      if (!this.ready) {
        wx.ready(() => {
          this.ready = true
          this.bus.emit('ready')
        })
        wx.error((err: any) => {
          this.bus.emit('error', err)
        })
      }
    }

    if (typeof wx !== 'undefined') {
      config()
    } else {
      const autoLoadJSSDK =
        params.autoLoadJSSDK == null ? '1.4.0' : params.autoLoadJSSDK
      if (autoLoadJSSDK !== false) {
        const jssdkUrl = /^[0-9.]+$/.test(autoLoadJSSDK)
          ? `https://res.wx.qq.com/open/js/jweixin-${autoLoadJSSDK}.js`
          : autoLoadJSSDK
        const alternateJssdkUrl = jssdkUrl.replace(
          /^https:\/\/res\.wx\.qq\.com\//,
          'https://res2.wx.qq.com/',
        )
        loadResource({
          type: LoadResourceUrlType.js,
          path: jssdkUrl,
          alternatePath: alternateJssdkUrl,
        })
          .then(([scriptEl]) => {
            scriptEl.parentNode?.removeChild(scriptEl)
          }, noop)
          .then(() => {
            if (typeof wx === 'undefined') {
              throw new Error('微信 JSSDK 加载失败')
            }
            config()
          })
      } else {
        if (typeof wx === 'undefined') {
          throw new Error('请先引入微信 JSSDK')
        }
        config()
      }
    }
  }

  /**
   * 判断当前客户端版本是否支持指定 JS 接口。
   *
   * @param jsApiList 需要检测的 JS 接口列表
   * @returns 以键值对的形式返回，可用的 `api` 值 `true`，不可用为 `false`
   */
  checkJsApi<T extends WechatJsApi>(
    jsApiList: T[],
  ): Promise<Record<T, boolean>> {
    return this.invoke<
      {
        jsApiList: T[]
      },
      {
        checkResult: Record<T, boolean>
      }
    >('checkJsApi', { jsApiList }).then(res => res.checkResult)
  }

  /**
   * 设置分享数据。
   *
   * **注意**：每次分享的数据会和上次分享的数据合并作为最终分享的数据，因此，可以设置全局的分享数据。
   *
   * @param params 分享数据
   */
  updateShareData(params: WechatUpdateShareDataParams): Promise<void> {
    params = {
      ...this.prevShareParams,
      ...params,
    }
    this.prevShareParams = params
    params = mapValues(params, value =>
      typeof value === 'function' ? value() : value,
    )
    return shareJsApiList.reduce<Promise<any>>((prev, jsApi) => {
      const next = () => this.invoke<WechatUpdateShareDataParams>(jsApi, params)
      return prev.then(next, next)
    }, Promise.resolve())
  }

  /**
   * 选择图片。
   *
   * @param params 参数
   * @returns 选定照片的本地 ID 列表，它们可以作为 img 标签的 src 属性显示图片
   */
  chooseImage(params?: WechatChooseImageParams): Promise<string[]> {
    return this.invoke<
      WechatChooseImageParams,
      {
        localIds: string[]
      }
    >('chooseImage', params).then(res => res.localIds)
  }

  /**
   * 预览图片。
   *
   * @param params 参数
   */
  previewImage(params: WechatPreviewImageParams): Promise<any> {
    return this.invoke<WechatPreviewImageParams>('previewImage', {
      urls: params.urls,
      current: params.current || params.urls[0],
    })
  }

  /**
   * 上传图片。
   *
   * **备注：** 上传图片有效期3天，
   * 可用微信多媒体接口下载图片到自己的服务器，
   * 此处获得的服务器端 ID 即 `media_id`。
   *
   * @param params 参数
   * @returns 图片的服务器端 ID
   */
  uploadImage(params: WechatUploadImageParams): Promise<string> {
    return this.invoke<
      {
        localId: string
        isShowProgressTips: 0 | 1
      },
      {
        serverId: string
      }
    >('uploadImage', {
      localId: params.localId,
      isShowProgressTips: params.isShowProgressTips ? 1 : 0,
    }).then(res => res.serverId)
  }

  /**
   * 获取地理位置接口。
   *
   * @param params 参数
   */
  getLocation(
    params?: WechatGetLocationParams,
  ): Promise<WechatGetLocationResult> {
    return this.invoke<WechatGetLocationParams>('getLocation', params || {})
  }

  /**
   * 使用微信内置地图查看位置。
   *
   * @param params 参数
   */
  openLocation(params: WechatOpenLocationParams): Promise<any> {
    return this.invoke<WechatOpenLocationParams>('openLocation', params)
  }

  /**
   * 关闭当前网页窗口。
   */
  closeWindow(): Promise<any> {
    return this.invoke('closeWindow')
  }

  /**
   * 批量隐藏非基础菜单项。
   *
   * @param menuList 要隐藏的非基础菜单项列表
   */
  hideNonBaseMenuItems(menuList: WechatNonBaseMenuItem[]): Promise<any> {
    return this.invoke<{
      menuList: WechatNonBaseMenuItem[]
    }>('hideMenuItems', { menuList })
  }

  /**
   * 批量显示非基础菜单项。
   *
   * @param menuList 要显示的非基础菜单项列表
   */
  showNonBaseMenuItems(menuList: WechatNonBaseMenuItem[]): Promise<any> {
    return this.invoke<{
      menuList: WechatNonBaseMenuItem[]
    }>('showMenuItems', { menuList })
  }

  /**
   * 隐藏所有的非基础菜单项。
   */
  hideAllNonBaseMenuItems(): Promise<any> {
    return this.invoke('hideAllNonBaseMenuItem')
  }

  /**
   * 显示所有的非基础菜单项。
   */
  showAllNonBaseMenuItems(): Promise<any> {
    return this.invoke('showAllNonBaseMenuItem')
  }

  /**
   * 发起微信支付。
   *
   * @param params 参数
   */
  requestPayment(params: WechatRequestPaymentParams): Promise<any> {
    return this.invoke('chooseWXPay', params)
  }

  /**
   * 错误处理。
   *
   * @param callback 出错时的回调函数
   */
  onError(callback: WechatErrorCallback): EventBusOffListener {
    return this.bus.on('error', callback)
  }

  /**
   * 调用 JSSDK 的 API 方法。
   *
   * @param jsApi 要调用的 API 名称
   * @param params 传给 API 的参数
   * @returns 调用结果
   */
  invoke<P extends Record<string, any> = Record<string, any>, T = any>(
    jsApi: WechatJsApi,
    params: P = {} as any,
  ): Promise<T> {
    return new Promise((resolve, reject) => {
      const invoke = () => {
        if (!wx[jsApi]) return reject(`wx.${jsApi} 不可用`)
        this.config()
        wx[jsApi]({
          ...params,
          success: resolve,
          fail: reject,
          cancel: reject,
        })
      }
      if (typeof wx === 'undefined' || !this.ready) {
        this.bus.once('ready', invoke)
      } else {
        invoke()
      }
    })
  }
}
