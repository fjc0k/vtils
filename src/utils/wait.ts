/**
 * @public
 */
export interface WaitResult<T> extends Promise<T> {
  /**
   * 取消等待，不执行后续逻辑。
   */
  cancel: () => void
}

/**
 * 等待一段时间 resolve。
 *
 * @public
 * @param milliseconds 等待时间(毫秒)
 * @param value resolve 值
 * @example
 * ```typescript
 * wait(1000).then(() => {
 *   console.log('ok')
 * }) // => 1秒后在控制台打印字符串: ok
 * ```
 */
export function wait<T>(milliseconds: number, value?: T): WaitResult<T> {
  let timer: any
  const result = new Promise<T | undefined>(resolve => {
    timer = setTimeout(() => resolve(value), milliseconds)
  }) as WaitResult<T>
  result.cancel = () => clearTimeout(timer)
  return result
}

/**
 * 等待一段时间后 reject。
 *
 * @public
 * @param milliseconds 等待时间(毫秒)
 * @param value reject 值
 * @example
 * ```typescript
 * wait.reject(1000).catch(() => {
 *   console.log('ok')
 * }) // => 1秒后在控制台打印字符串: ok
 * ```
 */
wait.reject = function reject(
  milliseconds: number,
  value?: any,
): WaitResult<never> {
  const waitRes = wait(milliseconds)
  const res: WaitResult<never> = waitRes.then(() =>
    Promise.reject(value),
  ) as any
  res.cancel = waitRes.cancel
  return res
}
