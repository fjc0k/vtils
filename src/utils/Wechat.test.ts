import { AnyFunction } from '../types'
import { wait } from './wait'
import { Wechat } from './Wechat'

describe('Wechat', () => {
  beforeAll(() => {
    let ready = false
    const readyCbs: AnyFunction[] = []
    const errorCbs: AnyFunction[] = []

    ;(window as any).wx = new Proxy(
      {},
      {
        get(_, method) {
          if (method === 'ready') {
            return (cb: AnyFunction) => {
              if (ready) return cb()
              if (readyCbs.length === 0) {
                wait(1000).then(() => {
                  ready = true
                  readyCbs.forEach(cb => cb())
                })
              }
              readyCbs.push(cb)
            }
          }

          if (method === 'error') {
            return (cb: AnyFunction) => {
              errorCbs.push(cb)
            }
          }

          return (params: { success: AnyFunction; fail: AnyFunction }) => {
            if (method === 'testFail') {
              return params.fail && params.fail()
            }
            if (method === 'testError') {
              errorCbs.forEach(cb => cb())
            }
            return params.success && params.success({})
          }
        },
      },
    )
  })

  const wechatConfigParams = {
    debug: false,
    appId: 'xxx',
    timestamp: 203093,
    nonceStr: 'dd',
    signature: 'xx',
    sharable: true,
    autoLoadJSSDK: '1.4.1',
  }

  test('可以正常创建实例', () => {
    const wechat = new Wechat()

    expect(wechat).toBeInstanceOf(Wechat)
    expect(wechat.configParams).toEqual({})
  })

  test('可以在创建实例时传入配置参数', () => {
    const wechat = new Wechat(wechatConfigParams)

    expect(wechat.configParams).toEqual(wechatConfigParams)
  })

  test('可以延后传入配置参数', () => {
    const wechat = new Wechat()
    wechat.config(wechatConfigParams)

    expect(wechat.configParams).toEqual(wechatConfigParams)
  })

  test('可以正常调用微信 JSSDK 的方法', async () => {
    const wechat = new Wechat(wechatConfigParams)

    await expect(wechat.invoke('addCard')).toResolve()
    await expect(wechat.invoke('chooseImage', {})).toResolve()
  })

  test('可以正常处理调用微信 JSSDK 的方法失败的情况', async () => {
    const wechat = new Wechat(wechatConfigParams)

    await expect(wechat.invoke('testFail' as any)).toReject()
  })

  test('可以收集错误信息', async () => {
    const wechat = new Wechat(wechatConfigParams)

    const errHandler0 = jest.fn()
    const errHandler1 = jest.fn()
    const errHandler2 = jest.fn()

    wechat.onError(errHandler0)
    wechat.onError(errHandler1)

    await wechat.invoke('testError' as any)

    wechat.onError(errHandler2)

    expect(errHandler0).toBeCalledTimes(1)
    expect(errHandler1).toBeCalledTimes(1)
    expect(errHandler2).toBeCalledTimes(0)
  })

  test('正常调用内置方法', async () => {
    const wechat = new Wechat(wechatConfigParams)

    await expect(wechat.checkJsApi(['chooseImage'])).toResolve()
    await expect(wechat.updateShareData({})).toResolve()
    await expect(wechat.chooseImage()).toResolve()
    await expect(wechat.previewImage({ urls: ['https://foo.bar'] })).toResolve()
    await expect(wechat.uploadImage({ localId: 'ss' })).toResolve()
    await expect(wechat.closeWindow()).toResolve()
    await expect(
      wechat.hideNonBaseMenuItems(['menuItem:share:brand']),
    ).toResolve()
    await expect(
      wechat.showNonBaseMenuItems(['menuItem:originPage']),
    ).toResolve()
    await expect(wechat.hideAllNonBaseMenuItems()).toResolve()
    await expect(wechat.showAllNonBaseMenuItems()).toResolve()
    await expect(
      wechat.openLocation({ latitude: 60, longitude: 40, name: '云南' }),
    ).toResolve()
  })

  test('beforeInvoke 正常', async () => {
    const wechat = new Wechat()
    const invokeHistory: any = []
    wechat.beforeInvoke((jsApi, params) => {
      invokeHistory.push({
        jsApi,
        params,
      })
    })
    wechat.config(wechatConfigParams)
    await wechat.updateShareData({})
    await wechat.closeWindow()
    expect(invokeHistory).toMatchSnapshot()
  })
})
