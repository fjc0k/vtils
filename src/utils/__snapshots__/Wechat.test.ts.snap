// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wechat beforeInvoke 正常 1`] = `
Array [
  Object {
    "jsApi": "updateAppMessageShareData",
    "params": Object {},
  },
  Object {
    "jsApi": "updateTimelineShareData",
    "params": Object {},
  },
  Object {
    "jsApi": "onMenuShareAppMessage",
    "params": Object {},
  },
  Object {
    "jsApi": "onMenuShareTimeline",
    "params": Object {},
  },
  Object {
    "jsApi": "onMenuShareQQ",
    "params": Object {},
  },
  Object {
    "jsApi": "onMenuShareQZone",
    "params": Object {},
  },
  Object {
    "jsApi": "closeWindow",
    "params": Object {},
  },
]
`;
